### PR TITLE
ddl: load table ranges from PD instead of region cache

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

Previously, `splitTableRanges()` uses region cache to load table ranges. But the cache entry may be outdate and invalid, we may get a old region that contains the whole data of the table. As a result, the performance of adding index can be bad.

### What changed and how does it work?

- Load table range from PD instead of region cache.
- Add log after `splitTableRange` to help debugging.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  image
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

**Review by D3Hunter — DISMISSED** [review]

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/ddl/backfilling.go [hunk 1]
```diff
@@ -427,79 +426,81 @@ func (w *backfillWorker) run(d *ddlCtx, bf backfiller, job *model.Job) {
...
+		return []kv.KeyRange{{StartKey: startKey, EndKey: endKey}}, nil
 	}
 
+	rc := s.GetRegionCache()
 	maxSleep := 10000 // ms
-	bo := backoff.NewBackofferWithVars(ctx, maxSleep, nil)
-	rc := copr.NewRegionCache(s.GetRegionCache())
-	ranges, err := rc.SplitRegionRanges(bo, []kv.KeyRange{kvRange}, limit)
+	bo := tikv.NewBackofferWithVars(ctx, maxSleep, nil)
+	rs, err := rc.BatchLoadRegionsWithKeyRange(bo, startKey, endKey, limit)
```
Comment (lance6716) [inline pkg/ddl/backfilling.go:457]: I check its logic with my lightning knowledge. In
BatchLoadRegionsWithKeyRange -> scanRegions -> handleRegionInfos (needLeader is `true`), seems if the region from PD does not contains leader store, the region will not return to caller. You have added a `validateAndFillRanges` function to find this case.

This defect of scan region API causes many duplicated logic, like lightning/BR's `checkRegionConsistency`. I have asked PD dev to move these checking logic into PD client.

lgtm for this PR, let's push PD dev @okJiang

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/backfilling.go [hunk 2]
```diff
@@ -427,79 +426,81 @@ func (w *backfillWorker) run(d *ddlCtx, bf backfiller, job *model.Job) {
...
+			}
+		}
+		if i == len(ranges)-1 {
+			e := r.EndKey
+			if len(e) == 0 || bytes.Compare(e, endKey) > 0 {
+				ranges[i].EndKey = endKey
+			}
+		}
+		if len(ranges[i].StartKey) == 0 || len(ranges[i].EndKey) == 0 {
+			return errors.Errorf("get empty start key in the middle of ranges")
```
Comment (tangenta) [inline pkg/ddl/backfilling.go:498]:
```suggestion
			return errors.Errorf("get empty start/end key in the middle of ranges")
```

**Review by lance6716 — COMMENTED** [review]
/hold

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**